test/unittests/test_dd_noise_functionality.cpp

```diff
@@ -0,0 +1,351 @@
+            dd,
+            qc.getNqubits(),
+            0.01,
+            0.02,
+            0.02,
+            0.04,
+            noiseEffects,
+            true,
+            false);
+
+    for (auto const& op: qc) {
+        auto operation = dd::getDD(op.get(), dd);
+        dd->applyOperationToDensity(rootEdge, operation, true);
+
+        deterministicNoiseFunctionality.applyNoiseEffects(rootEdge, op);
+    }
+
+    const auto m = dd->getProbVectorFromDensityMatrix(rootEdge, 0.001);
+
+    double tolerance = 1e-10;
+    EXPECT_NEAR(m.find("0000")->second, 0.0969332192741, tolerance);
+    EXPECT_NEAR(m.find("0001")->second, 0.0907888041538, tolerance);
+    EXPECT_NEAR(m.find("0010")->second, 0.0141409660985, tolerance);
+    EXPECT_NEAR(m.find("0100")->second, 0.0238203475524, tolerance);
+    EXPECT_NEAR(m.find("0101")->second, 0.0235097990017, tolerance);
+    EXPECT_NEAR(m.find("0110")->second, 0.0244576087400, tolerance);
+    EXPECT_NEAR(m.find("0111")->second, 0.0116282811276, tolerance);
+    EXPECT_NEAR(m.find("1000")->second, 0.1731941264570, tolerance);
+    EXPECT_NEAR(m.find("1001")->second, 0.4145855071998, tolerance);
+    EXPECT_NEAR(m.find("1010")->second, 0.0138062113213, tolerance);
+    EXPECT_NEAR(m.find("1011")->second, 0.0184033482066, tolerance);
+    EXPECT_NEAR(m.find("1100")->second, 0.0242454336917, tolerance);
+    EXPECT_NEAR(m.find("1101")->second, 0.0262779844799, tolerance);
+    EXPECT_NEAR(m.find("1110")->second, 0.0239296920989, tolerance);
+    EXPECT_NEAR(m.find("1111")->second, 0.0110373166627, tolerance);
+}
+
+TEST_F(DDNoiseFunctionalityTest, StochSimulateAdder4TrackAPD) {
+    auto dd = std::make_unique<StochasticNoiseTestPackage>(qc.getNqubits());
+
+    std::map<std::string, double, std::less<>> measSummary = {
+            {"0000", 0.},
+            {"0001", 0.},
+            {"0010", 0.},
+            {"0011", 0.},
+            {"0100", 0.},
+            {"0101", 0.},
+            {"0110", 0.},
+            {"0111", 0.},
+            {"1000", 0.},
+            {"1001", 0.},
+            {"1010", 0.},
+            {"1011", 0.},
+            {"1100", 0.},
+            {"1101", 0.}};
+
+    const auto noiseEffects = {dd::amplitudeDamping, dd::phaseFlip, dd::depolarization};
+
+    auto stochasticNoiseFunctionality = dd::StochasticNoiseFunctionality<StochasticNoiseTestPackage>(
+            dd,
+            qc.getNqubits(),
+            0.01,
+            0.02,
+            2.,
+            noiseEffects);
+
+    for (size_t i = 0U; i < stochRuns; i++) {
+        auto rootEdge = dd->makeZeroState(qc.getNqubits());
+        dd->incRef(rootEdge);
+
+        for (auto const& op: qc) {
+            auto        operation  = dd::getDD(op.get(), dd);
+            std::vector usedQubits = op->getTargets();
+            for (auto control: op->getControls()) {
+                usedQubits.push_back(control.qubit);
+            }
```

Comment:
code like this is used in several places. please extract this into a method of the `Operation` class and suitably override it for the individual operation types.